From 5c7133b4b00a97b3cfa20510782eb526687aeece Mon Sep 17 00:00:00 2001
From: Tong Tiangen <tongtiangen@huawei.com>
Date: Thu, 12 Jan 2023 11:22:01 +0000
Subject: [PATCH] kasan: add kasan support for memcpy_mcs()
Patch-mainline: Not yet, from openEuler
References: bsn#22
openEuler-commit: 5c7133b4b00a97b3cfa20510782eb526687aeece
Modified-by-SEL: No


hulk inclusion
category: bugfix
bugzilla: https://gitee.com/openeuler/kernel/issues/I6AQE8
CVE: NA

-------------------------------

The function memcpy_mcs() is designed to copy in memory and should be
supported by kasan.

Signed-off-by: Tong Tiangen <tongtiangen@huawei.com>
(cherry picked from commit 525213f1b1a5dac10ca2ffd5f2af4e28849ede5e)
Signed-off-by: Guoqing Jiang <guoqing.jiang@suse.com>
---
 mm/kasan/common.c | 12 ++++++++++++
 1 file changed, 12 insertions(+)

diff --git a/mm/kasan/common.c b/mm/kasan/common.c
index f8cd1e20b6c6..592eeba0a787 100644
--- a/mm/kasan/common.c
+++ b/mm/kasan/common.c
@@ -109,6 +109,18 @@ void *memcpy(void *dest, const void *src, size_t len)
 	return __memcpy(dest, src, len);
 }
 
+#ifdef __HAVE_ARCH_MEMCPY_MC
+#undef memcpy_mcs
+unsigned long memcpy_mcs(void *dest, const void *src, size_t len)
+{
+	if (!check_memory_region((unsigned long)src, len, false, _RET_IP_) ||
+	    !check_memory_region((unsigned long)dest, len, true, _RET_IP_))
+		return (unsigned long)len;
+
+	return __memcpy_mcs(dest, src, len);
+}
+#endif
+
 /*
  * Poisons the shadow memory for 'size' bytes starting from 'addr'.
  * Memory addresses should be aligned to KASAN_SHADOW_SCALE_SIZE.
-- 
2.33.0

